mailserver/cleaner.go

```diff
@@ -1,30 +1,91 @@
 package mailserver
 
 import (
+	"fmt"
+	"sync"
+	"time"
+
 	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
 	"github.com/syndtr/goleveldb/leveldb"
 	"github.com/syndtr/goleveldb/leveldb/iterator"
 	"github.com/syndtr/goleveldb/leveldb/util"
 )
 
-const batchSize = 1000
+const (
+	cleanerBatchSize = 1000
+	cleanerPeriod    = time.Hour
+)
+
+// cleaner removes old messages from a db.
+type cleaner struct {
+	sync.RWMutex
 
-// Cleaner removes old messages from a db
-type Cleaner struct {
 	db        dbImpl
 	batchSize int
+	retention time.Duration
+
+	period time.Duration
+	cancel chan struct{}
 }
 
-// NewCleanerWithDB returns a new Cleaner for db
-func NewCleanerWithDB(db dbImpl) *Cleaner {
-	return &Cleaner{
+// NewCleanerWithDB returns a new cleaner for db.
+func newCleanerWithDB(db dbImpl, retention time.Duration) *cleaner {
```

Comment:
`newClearnerForDB`?

Reply:
I have usually seen `WithSomething` pattern if a parameter is assigned to a field in a struct.

Reply:
yeah, my idea was that your comment says that it is a cleaner *for* a db, which is logical.
to me, having something *with* something means that it uses the db for it's own reasons.
maybe we can call it just `NewDBCleaner` then it is obvious that it should accept a db somewhere in params?

Reply:
That makes sense. `NewDBCleaner` seems like the most suitable option here.